File: src/FSharpVSPowerTools/Commands/LintTaggerProvider.cs
﻿using FSharpVSPowerTools.Linting;
using FSharpVSPowerTools.ProjectSystem;
using Microsoft.VisualStudio.Shell;
using Microsoft.VisualStudio.Text;
using Microsoft.VisualStudio.Text.Adornments;
using Microsoft.VisualStudio.Text.Classification;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Text.Tagging;
using Microsoft.VisualStudio.Utilities;
using System;
using System.ComponentModel.Composition;
using System.Windows.Media;

namespace FSharpVSPowerTools
{
    static class FormatDefinition
    {
        [Export(typeof(EditorFormatDefinition))]
        [Name(Constants.LintTagErrorType)]
        [Order(After = Priority.High)]
        [UserVisible(true)]
        internal class LintFormatDefinition : EditorFormatDefinition
        {
            public LintFormatDefinition()
            {
                this.ForegroundColor = Colors.Orange;
                this.BackgroundCustomizable = false;
                this.DisplayName = "F# Lint";
            }
        }

        [Export(typeof(ErrorTypeDefinition))]
        [Name(Constants.LintTagErrorType)]
        [DisplayName(Constants.LintTagErrorType)]
        internal static ErrorTypeDefinition LintErrorTypeDefinition = null;
    }

    [Export(typeof(IViewTaggerProvider))]
    [ContentType("F#")]
    [TagType(typeof(LintTag))]
    [TextViewRole(PredefinedTextViewRoles.Editable)]
    [TextViewRole(PredefinedTextViewRoles.PrimaryDocument)]
    public class LintTaggerProvider : IViewTaggerProvider
    {
        readonly IServiceProvider _serviceProvider;
        readonly ITextDocumentFactoryService _textDocumentFactoryService;
        readonly ProjectFactory _projectFactory;
        readonly VSLanguageService _fsharpVsLanguageService;
        readonly IOpenDocumentsTracker _openDocumentTracker;
    
        [ImportingConstructor]
        public LintTaggerProvider(
            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
            ITextDocumentFactoryService textDocumentFactoryService,
            ProjectFactory projectFactory,
            VSLanguageService fsharpVsLanguageService,
            IOpenDocumentsTracker openDocumentTracker)
        {
            _serviceProvider = serviceProvider;
            _textDocumentFactoryService = textDocumentFactoryService;
            _projectFactory = projectFactory;
            _fsharpVsLanguageService = fsharpVsLanguageService;
            _openDocumentTracker = openDocumentTracker;
        }

        public ITagger<T> CreateTagger<T>(ITextView textView, ITextBuffer buffer) where T : ITag
        {
            if (textView.TextBuffer != buffer) return null;

            var generalOptions = SettingsContext.GeneralOptions;
            if (generalOptions == null || !generalOptions.LinterEnabled) return null;

            ITextDocument doc;
            if (_textDocumentFactoryService.TryGetTextDocument(buffer, out doc))
            {
                return buffer.Properties.GetOrCreateSingletonProperty(
                    () => new LintTagger(doc, _fsharpVsLanguageService, _serviceProvider, _projectFactory, _openDocumentTracker) as ITagger<T>);
            }

            return null;
        }
    }
}

File: src/FSharpVSPowerTools.Pure/Commands/LintTaggerProvider.fs
module FSharpVSPowerTools.LintTaggerProvider

open FSharpVSPowerTools.Linting
open FSharpVSPowerTools.ProjectSystem
open Microsoft.VisualStudio.Shell
open Microsoft.VisualStudio.Text
open Microsoft.VisualStudio.Text.Adornments
open Microsoft.VisualStudio.Text.Classification
open Microsoft.VisualStudio.Text.Editor
open Microsoft.VisualStudio.Text.Tagging
open Microsoft.VisualStudio.Utilities
open System
open System.ComponentModel.Composition
open System.Windows.Media


[<Export (typeof<EditorFormatDefinition>)>]
[<Name (Constants.LintTagErrorType)>]
[<Order (After = Priority.High)>]
[<UserVisible true>]

type internal LintFormatDefinition () =
    inherit EditorFormatDefinition ()
    do  base.ForegroundColor <- Nullable Colors.Orange
        base.BackgroundCustomizable <- Nullable false
        base.DisplayName <- "F# Lint"

[<Export(typeof<ErrorTypeDefinition>)>]
[<Name(Constants.LintTagErrorType)>]
[<DisplayName(Constants.LintTagErrorType)>]
let internal  LintErrorTypeDefinition : ErrorTypeDefinition = null
    

[<Export (typeof<IViewTaggerProvider>)>]
[<ContentType "F#">]
[<TagType (typeof<LintTag>)>]
type LintTaggerProvider [<ImportingConstructor>]
    ( [<Import(typeof<SVsServiceProvider>)>] 
        serviceProvider                 : IServiceProvider              ,
        textDocumentFactoryService      :   ITextDocumentFactoryService ,
        projectFactory                  :   ProjectFactory              ,
        vsLanguageService               :   VSLanguageService           ) =

    interface IViewTaggerProvider with
        member __.CreateTagger (textView, buffer) =
            maybe{
                let! generalOptions = Setting.tryGetGeneralOptions serviceProvider
                if not generalOptions.LinterEnabled then return! None else
                let! doc = textDocumentFactoryService.TryDocumentFromBuffer buffer
                return buffer.Properties.GetOrCreateSingletonProperty (typeof<LintTagger>, fun () ->
                    new LintTagger 
                       (doc, vsLanguageService, serviceProvider, 
                        projectFactory, OpenDocumentsTracker textDocumentFactoryService)
                ) :> obj :?> _
            } |> Option.getOrElse null
